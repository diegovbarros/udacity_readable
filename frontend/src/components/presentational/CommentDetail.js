import React ,{Component, Fragment} from 'react'
import PropTypes from 'prop-types';
import {formatDate} from '../../utilities/helpers'
import LoadingCircular from '../presentational/LoadingCircular'

const propTypes = {
    comment : PropTypes.shape({
        id: PropTypes.string.isRequired,
        parentId : PropTypes.string.isRequired,
        timestamp: PropTypes.number.isRequired,
        body: PropTypes.string.isRequired,
        author: PropTypes.string.isRequired,
        votes: PropTypes.arrayOf(PropTypes.shape({
            user : PropTypes.string,
            option : PropTypes.string,
            value : PropTypes.number
        })).isRequired,
      }).isRequired,
      authUser : PropTypes.shape({
        name: PropTypes.string.isRequired,
        avatar : PropTypes.string.isRequired
      }).isRequired,
      onDeleteComment   : PropTypes.func.isRequired,
      onBindComment     : PropTypes.func.isRequired,
      onVoteComment     : PropTypes.func.isRequired,
      loading           : PropTypes.bool.isRequired
};


/**
* @description 
* Componente que representa um resumo do post dentro da lista de posts carregados
* @param {Object} comment              Comment
* @param {Object} authUser             Usuário Logado
* @param {Function} onDeleteComment    Método responsável por deletar o comment
* @param {Function} onBindComment      Método responsável por criar um comentário para adição ou atualização
* @param {Function} onVoteComment      Método responsável pelo votescore do comentário
* @param {Boolean} loading             Se é necessário mostrar o loading

*/
class CommentDetail extends Component  {

    state = {
        showLoading: false,
    };
    
    componentWillUnmount() {
        this.setState({showLoading : false});
    }

    onDelete = (id,event) => {
        this.setState({showLoading : true})
        const {onDeleteComment} = this.props
        onDeleteComment(id,event)
    }

    setClassName = (option) =>{
        const {comment,authUser}    = this.props
        const vote                  = comment.votes.find((v) => v.user === authUser.name && v.option === option)
        return (vote) ? "active" : ""
    }

    render() {
        const {comment,authUser,onBindComment,onVoteComment} = this.props
        const {showLoading}     = this.state
        return (
            <li className="comment-item">
                {showLoading && <LoadingCircular></LoadingCircular>}
                {!showLoading &&
                    <Fragment>
                        <div className="post__author author vcard inline-items">
                            <div className="author-date">
                                <a className="h6 post__author-name fn" href="02-ProfilePage.html">{comment.author}</a>
                                <div className="post__date">
                                    <time className="published" dateTime={formatDate(comment.timestamp)}>
                                        {formatDate(comment.timestamp)}
                                    </time>
                                </div>
                            </div>
                            {comment.author === authUser.name && 
                                <div className="more">
                                    <svg className="olymp-three-dots-icon">
                                        <use xlinkHref="/img/svg-icons/sprites/icons.svg#olymp-three-dots-icon"></use>
                                    </svg>
                                    <ul className="more-dropdown">
                                        <li>
                                            <a href="" onClick={(event) => onBindComment(comment,event)}>Edit Post</a>
                                        </li>
                                        <li>
                                            <a href="" onClick={(event) => this.onDelete(comment.id,event)}>Delete Post</a>
                                        </li>
                                    </ul>
                                </div>
                            }
                        </div>
                        <p>{comment.body}</p>
                            <a href="" onClick={(event) => onVoteComment(comment.id,authUser.name,"upVote",event)} className={`post-add-icon inline-items ${this.setClassName("downVote")}`}>
                                <svg className="olymp-heart-icon">
                                    <title>Up</title>
                                    <use xlinkHref="/img/svg-icons/sprites/icons.svg#olymp-heart-icon"></use>
                                </svg>
                            </a>
                            <a href="" onClick={(event) => onVoteComment(comment.id,authUser.name,"downVote",event) } className={`post-add-icon inline-items ${this.setClassName("downVote")}`}>
                                <svg id="olymp-heart-icon" viewBox="0 0 36 32" width="100%" height="100%">
                                    <title>Down</title>
                                    <g transform="translate(35,30) rotate(180)">
                                        <use xlinkHref="/img/svg-icons/sprites/icons.svg#olymp-heart-icon"></use>
                                    </g>
                                </svg>
                            </a>
                            <span>
                                {comment.votes.reduce((acc,vote) => {
                                        return acc + vote.value
                                    },0)
                                }
                            </span>
                    </Fragment>
                    
                }
           </li>
        )
    }
}

CommentDetail.propTypes    = propTypes;

export default CommentDetail
