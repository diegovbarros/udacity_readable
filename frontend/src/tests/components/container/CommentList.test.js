import React from 'react';
import ReactRouterEnzymeContext from 'react-router-enzyme-context';
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import CommentListConnect ,{CommentList}from '../../../components/container/CommentList';
import CommentDetail from '../../../components/presentational/CommentDetail';

import commentsMock from '../../mocks/comments';

describe('[Container Component] CommentList', () => {
  const context 	= new ReactRouterEnzymeContext();
	const comments 	= Object.keys(commentsMock).map((id) => commentsMock[id])
	let props 			= {}

	beforeEach(() => {
		props = {
			comments,
			authUser : {name: "Udacity",avatar: "udacity"},
			onDeleteComment: jest.fn(),
			voteComment: jest.fn(),
			getComments: jest.fn(),
			loading: false,
		};
	})

  it('Shallow renders correctly', () => {
    expect(global.shallow(<CommentList {...props} />,context.get()));
	});

	it('Shallow renders correctly', () => {

		const mockStore = configureMockStore([thunk]);
		const store = mockStore({
			user: {name:"Udacity",avatar: "udacity"}
		});

		expect(global.shallow(<CommentListConnect store={store} postId={"sa"} />,context.get()));
	});

	it('Renders correctly', () => {
    expect(global.shallow(<CommentList {...props} />)).toMatchSnapshot();
	});

	it('Shallow renders correctly', () => {

		const mockStore = configureMockStore([thunk]);
		const store = mockStore({
			user: {name:"Udacity",avatar: "udacity"}
		});

		expect(global.shallow(<CommentListConnect store={store} {...props} />)).toMatchSnapshot();
	});

	it('Expect to find same number os Comment Component equals to comments', () => {
    const wrapper = global.shallow(<CommentList {...props}/>);
    expect(wrapper.find(CommentDetail)).toHaveLength(props.comments.length);
	});

	it('Expect to not find Comment Component', () => {

		props = {
			...props,comments :[]
		}

    const wrapper = global.shallow(<CommentList {...props} />);
		expect(wrapper.find(CommentDetail)).toHaveLength(props.comments.length);
	});

	it('Expect to find label of comments quantity', () => {

		const wrapper = global.mount(<CommentList {...props} />, context.get());
    expect(wrapper.find('.heading-title').text()).toBeDefined();
    expect(wrapper.find('.heading-title').text()).not.toBeNull();
		expect(wrapper.find('.heading-title').text()).toEqual(`Comments (${props.comments.length})`);
	});


});
