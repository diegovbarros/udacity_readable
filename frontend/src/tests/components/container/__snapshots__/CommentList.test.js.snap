// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Container Component] CommentList Renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CommentList
    authUser={
        Object {
            "avatar": "udacity",
            "name": "Udacity",
          }
    }
    comments={
        Array [
            Object {
              "author": "Udacity",
              "body": "Hi there! I am a COMMENT.",
              "deleted": false,
              "id": "894tuq4ut84ut8v4t8wun89g",
              "parentDeleted": false,
              "parentId": "8xf0y6ziyjabvozdd253n",
              "timestamp": 1468166872634,
              "votes": Array [],
            },
            Object {
              "author": "Udacity",
              "body": "Comments. Are. Cool.",
              "deleted": false,
              "id": "8tu4bsun805n8un48ve89",
              "parentDeleted": false,
              "parentId": "8xf0y6ziyjabvozdd253nd",
              "timestamp": 1469479767190,
              "votes": Array [],
            },
          ]
    }
    getComments={[Function]}
    loading={false}
    onDeleteComment={[Function]}
    onHandleSetComment={[Function]}
    postId=""
    voteComment={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="crumina-module crumina-heading with-title-decoration"
>
          <h5
                    className="heading-title"
          >
                    Comments (
                    2
                    )
          </h5>
</div>,
        <ul
          className="comments-list"
>
          <CommentDetail
                    authUser={
                              Object {
                                        "avatar": "udacity",
                                        "name": "Udacity",
                                      }
                    }
                    comment={
                              Object {
                                        "author": "Udacity",
                                        "body": "Hi there! I am a COMMENT.",
                                        "deleted": false,
                                        "id": "894tuq4ut84ut8v4t8wun89g",
                                        "parentDeleted": false,
                                        "parentId": "8xf0y6ziyjabvozdd253n",
                                        "timestamp": 1468166872634,
                                        "votes": Array [],
                                      }
                    }
                    loading={false}
                    onBindComment={[Function]}
                    onDeleteComment={[Function]}
                    onVoteComment={[Function]}
          />
          <CommentDetail
                    authUser={
                              Object {
                                        "avatar": "udacity",
                                        "name": "Udacity",
                                      }
                    }
                    comment={
                              Object {
                                        "author": "Udacity",
                                        "body": "Comments. Are. Cool.",
                                        "deleted": false,
                                        "id": "8tu4bsun805n8un48ve89",
                                        "parentDeleted": false,
                                        "parentId": "8xf0y6ziyjabvozdd253nd",
                                        "timestamp": 1469479767190,
                                        "votes": Array [],
                                      }
                    }
                    loading={false}
                    onBindComment={[Function]}
                    onDeleteComment={[Function]}
                    onVoteComment={[Function]}
          />
</ul>,
      ],
      "className": "ui-block",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h5
            className="heading-title"
>
            Comments (
            2
            )
</h5>,
          "className": "crumina-module crumina-heading with-title-decoration",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Comments (",
              2,
              ")",
            ],
            "className": "heading-title",
          },
          "ref": null,
          "rendered": Array [
            "Comments (",
            2,
            ")",
          ],
          "type": "h5",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <CommentDetail
              authUser={
                            Object {
                                          "avatar": "udacity",
                                          "name": "Udacity",
                                        }
              }
              comment={
                            Object {
                                          "author": "Udacity",
                                          "body": "Hi there! I am a COMMENT.",
                                          "deleted": false,
                                          "id": "894tuq4ut84ut8v4t8wun89g",
                                          "parentDeleted": false,
                                          "parentId": "8xf0y6ziyjabvozdd253n",
                                          "timestamp": 1468166872634,
                                          "votes": Array [],
                                        }
              }
              loading={false}
              onBindComment={[Function]}
              onDeleteComment={[Function]}
              onVoteComment={[Function]}
/>,
            <CommentDetail
              authUser={
                            Object {
                                          "avatar": "udacity",
                                          "name": "Udacity",
                                        }
              }
              comment={
                            Object {
                                          "author": "Udacity",
                                          "body": "Comments. Are. Cool.",
                                          "deleted": false,
                                          "id": "8tu4bsun805n8un48ve89",
                                          "parentDeleted": false,
                                          "parentId": "8xf0y6ziyjabvozdd253nd",
                                          "timestamp": 1469479767190,
                                          "votes": Array [],
                                        }
              }
              loading={false}
              onBindComment={[Function]}
              onDeleteComment={[Function]}
              onVoteComment={[Function]}
/>,
          ],
          "className": "comments-list",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "894tuq4ut84ut8v4t8wun89g",
            "nodeType": "class",
            "props": Object {
              "authUser": Object {
                "avatar": "udacity",
                "name": "Udacity",
              },
              "comment": Object {
                "author": "Udacity",
                "body": "Hi there! I am a COMMENT.",
                "deleted": false,
                "id": "894tuq4ut84ut8v4t8wun89g",
                "parentDeleted": false,
                "parentId": "8xf0y6ziyjabvozdd253n",
                "timestamp": 1468166872634,
                "votes": Array [],
              },
              "loading": false,
              "onBindComment": [Function],
              "onDeleteComment": [Function],
              "onVoteComment": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "8tu4bsun805n8un48ve89",
            "nodeType": "class",
            "props": Object {
              "authUser": Object {
                "avatar": "udacity",
                "name": "Udacity",
              },
              "comment": Object {
                "author": "Udacity",
                "body": "Comments. Are. Cool.",
                "deleted": false,
                "id": "8tu4bsun805n8un48ve89",
                "parentDeleted": false,
                "parentId": "8xf0y6ziyjabvozdd253nd",
                "timestamp": 1469479767190,
                "votes": Array [],
              },
              "loading": false,
              "onBindComment": [Function],
              "onDeleteComment": [Function],
              "onVoteComment": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="crumina-module crumina-heading with-title-decoration"
>
            <h5
                        className="heading-title"
            >
                        Comments (
                        2
                        )
            </h5>
</div>,
          <ul
            className="comments-list"
>
            <CommentDetail
                        authUser={
                                    Object {
                                                "avatar": "udacity",
                                                "name": "Udacity",
                                              }
                        }
                        comment={
                                    Object {
                                                "author": "Udacity",
                                                "body": "Hi there! I am a COMMENT.",
                                                "deleted": false,
                                                "id": "894tuq4ut84ut8v4t8wun89g",
                                                "parentDeleted": false,
                                                "parentId": "8xf0y6ziyjabvozdd253n",
                                                "timestamp": 1468166872634,
                                                "votes": Array [],
                                              }
                        }
                        loading={false}
                        onBindComment={[Function]}
                        onDeleteComment={[Function]}
                        onVoteComment={[Function]}
            />
            <CommentDetail
                        authUser={
                                    Object {
                                                "avatar": "udacity",
                                                "name": "Udacity",
                                              }
                        }
                        comment={
                                    Object {
                                                "author": "Udacity",
                                                "body": "Comments. Are. Cool.",
                                                "deleted": false,
                                                "id": "8tu4bsun805n8un48ve89",
                                                "parentDeleted": false,
                                                "parentId": "8xf0y6ziyjabvozdd253nd",
                                                "timestamp": 1469479767190,
                                                "votes": Array [],
                                              }
                        }
                        loading={false}
                        onBindComment={[Function]}
                        onDeleteComment={[Function]}
                        onVoteComment={[Function]}
            />
</ul>,
        ],
        "className": "ui-block",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h5
              className="heading-title"
>
              Comments (
              2
              )
</h5>,
            "className": "crumina-module crumina-heading with-title-decoration",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Comments (",
                2,
                ")",
              ],
              "className": "heading-title",
            },
            "ref": null,
            "rendered": Array [
              "Comments (",
              2,
              ")",
            ],
            "type": "h5",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <CommentDetail
                authUser={
                                Object {
                                                "avatar": "udacity",
                                                "name": "Udacity",
                                              }
                }
                comment={
                                Object {
                                                "author": "Udacity",
                                                "body": "Hi there! I am a COMMENT.",
                                                "deleted": false,
                                                "id": "894tuq4ut84ut8v4t8wun89g",
                                                "parentDeleted": false,
                                                "parentId": "8xf0y6ziyjabvozdd253n",
                                                "timestamp": 1468166872634,
                                                "votes": Array [],
                                              }
                }
                loading={false}
                onBindComment={[Function]}
                onDeleteComment={[Function]}
                onVoteComment={[Function]}
/>,
              <CommentDetail
                authUser={
                                Object {
                                                "avatar": "udacity",
                                                "name": "Udacity",
                                              }
                }
                comment={
                                Object {
                                                "author": "Udacity",
                                                "body": "Comments. Are. Cool.",
                                                "deleted": false,
                                                "id": "8tu4bsun805n8un48ve89",
                                                "parentDeleted": false,
                                                "parentId": "8xf0y6ziyjabvozdd253nd",
                                                "timestamp": 1469479767190,
                                                "votes": Array [],
                                              }
                }
                loading={false}
                onBindComment={[Function]}
                onDeleteComment={[Function]}
                onVoteComment={[Function]}
/>,
            ],
            "className": "comments-list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "894tuq4ut84ut8v4t8wun89g",
              "nodeType": "class",
              "props": Object {
                "authUser": Object {
                  "avatar": "udacity",
                  "name": "Udacity",
                },
                "comment": Object {
                  "author": "Udacity",
                  "body": "Hi there! I am a COMMENT.",
                  "deleted": false,
                  "id": "894tuq4ut84ut8v4t8wun89g",
                  "parentDeleted": false,
                  "parentId": "8xf0y6ziyjabvozdd253n",
                  "timestamp": 1468166872634,
                  "votes": Array [],
                },
                "loading": false,
                "onBindComment": [Function],
                "onDeleteComment": [Function],
                "onVoteComment": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "8tu4bsun805n8un48ve89",
              "nodeType": "class",
              "props": Object {
                "authUser": Object {
                  "avatar": "udacity",
                  "name": "Udacity",
                },
                "comment": Object {
                  "author": "Udacity",
                  "body": "Comments. Are. Cool.",
                  "deleted": false,
                  "id": "8tu4bsun805n8un48ve89",
                  "parentDeleted": false,
                  "parentId": "8xf0y6ziyjabvozdd253nd",
                  "timestamp": 1469479767190,
                  "votes": Array [],
                },
                "loading": false,
                "onBindComment": [Function],
                "onDeleteComment": [Function],
                "onVoteComment": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "ul",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
